src/AnalysisEpic.cxx

```diff
@@ -0,0 +1,424 @@
+      
+      else if(genStat_ == 1 && mcidmap[mcpart_.mcID]>-1){ 
+	kinTrue->AddToHFS(mcpart_.vecPart);
+      }
+    
+    }
+
+    //check beam finding
+    if(!foundBeamElectron || !foundBeamIon) { numNoBeam++; continue;};
+
+    /*
+      Loop over RecoParticles
+    */
+
+    int itrack = 0;
+    bool recEleFound=false;
+    for(auto trackpart_ : trackpart){
+      // Skip if there is no matching MCParticle
+      if(trackidmap[itrack]==-1) continue;
+      // If the trackidmap is linked to the genEleID (generated scattered electron), identify this reco particle as the electron
+      if(trackidmap[itrack]==genEleID){	
+	recEleFound=true;
+	kin->vecElectron= trackpart_.vecPart;
+      }
+      // Add the final state particle to the HFS
+      kin->AddToHFS(trackpart_.vecPart);
+      itrack++;
+    }
+
+    // Skip event if the reco scattered electron was missing
+    if(recEleFound==false){
+      numNoEle++;
+      continue;
+    }
+    else
+      numEle++;
+
+    // subtrct electron from hadronic final state variables
+    kin->SubtractElectronFromHFS();
+    kinTrue->SubtractElectronFromHFS();
+
+    // skip the event if there are no reconstructed particles (other than the
+    // electron), otherwise hadronic recon methods will fail
+    if(kin->countHadrons == 0) {
+      numNoHadrons++;
+      continue;
+    };
+   
+    // calculate DIS kinematics
+    if(!(kin->CalculateDIS(reconMethod))) continue; // reconstructed
+    if(!(kinTrue->CalculateDIS(reconMethod))) continue; // generated (truth)
+
+    // Get the weight for this event's Q2
+    auto Q2weightFactor = GetEventQ2Weight(kinTrue->Q2, inLookup[chain->GetTreeNumber()]);
+
+    // fill inclusive histograms, if only `inclusive` is included in output
+    // (otherwise they will be filled in track and jet loops)
+    if(includeOutputSet["inclusive_only"]) {
+      auto wInclusive = Q2weightFactor * weightInclusive->GetWeight(*kinTrue);
+      wInclusiveTotal += wInclusive;
+      FillHistosInclusive(wInclusive);
+    }
+
+    /*
+      Loop again over the reconstructed particles
+      Calculate Hadron Kinematics
+      Fill output data structures (Histos, SimpleTree, etc.)
+    */
+
+    for(auto part : trackpart){
+
+      int pid_ = part.pid;
+      int mcid_ = part.mcID;
+
+      // final state cut
+      // - check PID, to see if it's a final state we're interested in for
+      //   histograms; if not, proceed to next track
+      auto kv = PIDtoFinalState.find(pid_);
+      if(kv!=PIDtoFinalState.end()) finalStateID = kv->second; else continue;
+      if(activeFinalStates.find(finalStateID)==activeFinalStates.end()) continue;
+
+      // calculate reconstructed hadron kinematics
+      kin->vecHadron = part.vecPart;
+      kin->CalculateHadronKinematics();
+
+      // find the matching truth hadron using mcID, and calculate its kinematics
+      if(mcid_ > 0) {
+	for(auto imc : mcpart) {
+	  if(mcid_ == imc.mcID) {
+	    kinTrue->vecHadron = imc.vecPart;
+	    break;
```

Comment:
is it possible that `kinTrue->vecHadron` is not set for a given `part`?

Reply:
Yes, I believe that may happen if the mcid_ was 0, i.e., the Monte Carlo particles that matches with the given REC particle has an mcID equal to 0. The mcID of the MC Particles is their index in the MCParticle TBranches. To be safe, changing the if-statement a few lines above.